[Sean Koh Hak Guan] iP

sk2001git opened the pull request:
### Duke Pro 

> "Benedict Cucumber batch" - Some cucumber ([Source](https://en.wikipedia.org/wiki/Benedict_Cumberbatch))

**Duke Pro** is

-  A **student led project**

- A chatbot

- **Easy** to learn and use & _Fast_ ~Bogosort~ 👀 👀 

To download it, 

1. Download the JAR file from the repo 
2. Open your CLI and the jar using java --jar 
3. Type in text to the CLI to use it

If you are using Java, here is the `main` in Java method to run it 

```java 
public static void main(String[] args) {
        new Duke("./data/duke.txt").run();
}
```

Features: 
- [x] Managing task
- [ ] Managing deadlines (coming soon)
- [ ] Reminders (coming soon)

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

sk2001git committed:
Duke Level 0 - Rename, Greet and Edit

sk2001git committed:
Duke now Echos - allowing user input through scanner, echoing replies unless bye is written

sk2001git committed:
Added a storage to store words that are inputted. Has a limit of 100

sk2001git committed:
Added a Task class to encapsulate a Task event. Amended Duke accordingly to include marking and unmarking

sk2001git committed:
Using enums instead of if else statements in Duke. Enum class commands helps to discern between the different key commands

sk2001git committed:
Changed Enum slightly to use start of word suffix instead of contains. Additionally completed Level 4 by implementing the design as mentioned

sk2001git committed:
Automated Text UI Testing. Succesful Test using I/O technique

sk2001git committed:
Added exception handling if wrong inputs / commands are given.

sk2001git committed:
Change storage structure from Array to Collection ArrayList, amended the code that way, added taskDeleted to task to show right message.

sk2001git committed:
Updated imports to match exact packages by java util. Following coding standard

sk2001git committed:
Done feature for saving tasks into hard drive

Made use of File, File Writer. Created Storage Class, Task List Class

Additional OOP improvements to code as well

sk2001git committed:
Updated gitignore to ignore text file used for storage

sk2001git committed:
Merge branch 'branch-Level-7'

sk2001git committed:
Fix Bugs on Task List Parser and Printing statements

sk2001git committed:
Done the feature of incorporating java date time for deadlines
and events

sk2001git committed:
Merge branch 'branch-Level-8'

sk2001git committed:
Done more OOP by encapsulating the UI messages into 1 class,
Added a Parser class to abstract the logic from main chatbot

sk2001git committed:
Merge branch 'branch-A-MoreOOP'

sk2001git committed:
Created packages, all Java files are now under package duke

sk2001git committed:
Merge branch 'branch-A-Packages'

sk2001git committed:
Merge branch 'add-gradle-support'

sk2001git committed:
Set up gradle automation in file. ./gradlew run is able to work
Was missing settings.gradle.kts, had to cover over

sk2001git committed:
Added JUnitTesting for 2 classes 2 non-trival methods
Added catch for null case for parsing for event (Missed out)

sk2001git committed:
Merge branch 'branch-A-JUnit'

sk2001git committed:
Updated gradle for JAR building, deleted unnecessary code for
tasklistTester

sk2001git committed:
Added Java docs to testing classes as well as running classes

sk2001git committed:
Merge branch 'branch-A-JavaDocs'

sk2001git committed:
Updated Format and spacing to better represent coding standards

sk2001git committed:
Merge branch 'branch-A-CodingStandard'
Fixed Conflicts

# Conflicts:
#	src/main/java/duke/Command.java
#	src/main/java/duke/InvalidInputException.java
#	src/main/java/duke/Parser.java
#	src/main/java/duke/WrongInputTask.java

sk2001git committed:
Added Search Feature to search for any task in list

sk2001git committed:
Merge branch 'branch-Level-9'
Conflicts Resolved

# Conflicts:
#	src/main/java/duke/InvalidInputException.java
#	src/main/java/duke/Parser.java
#	src/main/java/duke/Task.java
#	src/main/java/duke/TaskList.java
#	src/main/java/duke/Ui.java

sk2001git committed:
Updated Java Coding Standard again, added more documentation

sk2001git committed:
Edited so it creates the directory if it does not exist

sk2001git committed:
Updated build gradle for JavaFx

sk2001git committed:
Incorporated Checkstyle of addressbook into current project

sk2001git committed:
Added CheckStyle and corrected violations

sk2001git committed:
Merge branch 'branch-A-CheckStyle'

sk2001git committed:
Trial setup using the JavaFX guide

sk2001git committed:
Separated the GUI, created a MVC to communicate between chatbot and gui

sk2001git committed:
Merge branch 'branch-Level-10'

PetrichorPrecipice commented on src/main/java/duke/ConsoleOutputStream.java:
```diff
@@ -0,0 +1,38 @@
+
+    /**
+     * Returns the content of the OutputStream as a String.
+     * @return  the content of the OutputStream as a String.
```
```suggestion
     * @return the content of the OutputStream as a String.
```

sk2001git replied:
Thank you, resolved!

PetrichorPrecipice commented on src/main/java/duke/Command.java:
```diff
@@ -0,0 +1,58 @@
+    /**
+     * This method is used to get the command from the user's input
+     * @param input the user's input
+     * @return  the command
```
```suggestion
     * @return the command
```

sk2001git replied:
Thanks!

PetrichorPrecipice commented on src/main/java/duke/ConsoleOutputStream.java:
```diff
@@ -0,0 +1,38 @@
+
+    /**
+     * Writes the specified byte to this output stream.
+     * @param b  the byte to be written.
```
```suggestion
     * @param b the byte to be written.
```

PetrichorPrecipice commented on src/main/java/duke/ConsoleOutputStream.java:
```diff
@@ -0,0 +1,38 @@
+
+    /**
+     * Returns the content of the OutputStream as a String.
+     * @return  the content of the OutputStream as a String.
```
```suggestion
     * @return the content of the OutputStream as a String.
```

PetrichorPrecipice commented on src/main/java/duke/DateTimeParser.java:
```diff
@@ -0,0 +1,67 @@
+ * This class is used to parse the date and time from the user input.
+ */
+public class DateTimeParser {
+    // Note that we take "MMM-dd-yyyy HHmm a" and "MMM-d-yyyy HHmm a" as valid date and time formats because
```
I like that you took the time to explain why you took it as a valid date!

PetrichorPrecipice commented on src/main/java/duke/DateTimeParser.java:
```diff
@@ -0,0 +1,67 @@
+    /**
+     * Creates a LocalDateTime object based on the input
+     * @param validDateTime a valid Date Time formatted String already checked by isValidDateTime
+     * @return  a LocalDateTime object
```
```suggestion
     * @return a LocalDateTime object
```
Should there be an extra space here?
I noticed a lot of your Javadocs for methods have this issue where you have an extra space following the @ return

sk2001git replied:
For documentation, I have this problem where i double space a lot, thanks for letting me know though, I will be more conscious !

PetrichorPrecipice commented on src/main/java/duke/Event.java:
```diff
@@ -0,0 +1,107 @@
+     * @param taskName  name of task.
+     * @param isDone    whether the task is done or not
+     * @param start    the start time of the event
+     * @param end    the end time of the event
```
```suggestion
     * @param taskName name of task.
     * @param isDone whether the task is done or not
     * @param start the start time of the event
     * @param end the end time of the event
```
Are these extra spaces supposed to be here?

PetrichorPrecipice commented on src/main/java/duke/FileManager.java:
```diff
@@ -0,0 +1,48 @@
+
+/**
+ * This class is used to represent the Management of the files
+ * TODO: Remove this class it is not needed
```
I like how you reminded yourself here to remove this!

PetrichorPrecipice commented on src/main/java/duke/DateTimeParser.java:
```diff
@@ -0,0 +1,67 @@
+    public static final String[] VALID_DATE_TIME_FORMAT_OUTPUT = {"MMM-dd-yyyy HHmm a", "MMM-d-yyyy HHmm a"};
+
+    /**
+     * Checks if the input is a valid date and time format
```
```suggestion
     * Checks if the input is a valid date and time format.
```
Should you place a full stop at the end here?
I noticed the same issue in other places too.

PetrichorPrecipice commented on src/main/java/duke/SearchEngine.java:
```diff
@@ -0,0 +1,54 @@
+    /**
+     * Searches the taskList for tasks that contain the input
+     * @param input
+     * @return
```
```suggestion
     * @return The TaskList containing all found entries.
```
Should this be left blank?

PetrichorPrecipice reviewed (commented):
LGTM! Overall very easy to read with the help of the comments. Just a few small coding standard violations and some spacing issues in regards to Javadocs. Otherwise looks great!

sk2001git committed:
Added Welcome Message to GUI

sk2001git committed:
Merge branch 'branch-WelcomeMessageFix'

czhiruo commented on src/main/java/duke/Deadline.java:
```diff
@@ -0,0 +1,90 @@
+
+    /**
+     * Converts a Deadline task into a string that can be saved
+     * @return  a string that can be saved
```
instead of `* @return  a string that can be saved` should be `* @return a string that can be saved` 
there shouldn't be an extra space in the javadocs

czhiruo commented on src/main/java/duke/Task.java:
```diff
@@ -0,0 +1,107 @@
+    public void completeTask() {
+        this.isDone = true;
+        System.out.println("Nice! I've marked this task as done:");
+        // Double space to maintain consistency with rubrics
```
I like how you explained this empty string

czhiruo commented on src/main/java/duke/Task.java:
```diff
@@ -0,0 +1,107 @@
+    }
+
+    /**
+     * Checks whether a task is completed and return a nX if done
```
```suggestion
     * Checks whether a task is completed and return a X if done.
```

sk2001git replied:
Thanks for the comments!

czhiruo commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,140 @@
+    /**
+     * Constructor for creating a Storage object
+     * @param filePath  the path of the file
+     * @param file  the file
```
```suggestion
     * @param file the file
```

czhiruo commented on src/main/java/duke/WrongInputException.java:
```diff
@@ -0,0 +1,16 @@
+    /**
+     * Constructor for creating a WrongInputTask exception
+     * @param taskExceptionMessage What is the error or invalid format
+     * @param solution  What is the recommended solution
```
```suggestion
     * @param solution The recommended solution
```

czhiruo replied:
Phrasing seems more like a question rather than a statement you could consider rephrasing this

czhiruo reviewed (commented):
code is well abstracted! perhaps slight styling changes to your documentation would be appreciated for greater readability.

sk2001git committed:
Updated gradle to reflect jar as well as duke main function to launch
gui

sk2001git committed:
Add assert to stop execution of programme given file corruption

There are no remedial action for default statements used under switch
statements

Amended the classes above in the commit: DateTime, Parser, TaskList
In DateTime:
* If we are to create a DateTime from storage and a WrongInputException
is thrown, that means that there is data corruption in the file storage
That is a invalid file format was stored. Thus we use assert to
terminate the program since the file is no longer safe to work with.

In Parser:
* Upon hitting the default case which means after parsing a command,
it did not throw an exception however it is invalid means that the
system has failed to detect an invalid command, the program should
terminate to prevent any further actions

In TaskList:
* We assert to check if the storage data is corrupted, that is if there
is even the minimum amount of data to form a task (Task name, status
and type). Programme should terminate to prevent wrong exception
message for the problem
* Assert in default statement since it is a invalid task however it was
created and stored , data is corrupted, should stop execution of program
* Assert in creating task List because we return null in ParseTask
This results in problematic null pointer exception that
could possibly be caught and given the wrong description.
Program is no longer functioning as it should be, should terminate

The asserts are done as a way to deal with highly unlikely but
theoretically possible situations that could cause multiple exceptions
and code functionality to stop working as intended. The strategy to
recover from this situation is to terminate since clearing data file
should not be called by default

sk2001git commented:
Amended code to improve code quality in the code violations spotted above

sk2001git committed:
Clean up code to better reflect code quality

Had consistency issue with guard clauses (some methods implementing
guard clauses while other using if for happy path)

Amended them to have guard clauses and no default branches are used,
that is else statements are removed, changed to if else or if statements
as guard clauses.

Additionally, there was unused methods, parameters in constructors and
classes

Remove File Manager class which no methods were implemented, changed
constructors of several classes like Deadline, Event, SearchEngine
where parameters were used, removing the redundant fields as well

This was done to clean up code that was unused

Apply DRY to prevent code duplication. Code that was defined
twice was condensed, this was seen in task Validator in deadline and
Event class

Apply SLAP in Parser class to achieve single level abstraction as well
as same abstraction levels

Magic Number like offsets to determine taskName for ToDos, Deadline,
Events were declared as final static constants in their respective class

This increases readability in terms of how we are searching for taskName

Deep Nesting was also checked for, there is no loops that is more than
2 nest deep as a general guideline

Long methods are kept minimal, even though impossible to keep to a hard
constraint of 30 lines for some methods like parsing

Code quality in total was cleaned up for all classes and scanned through
Code remains functional after changes as well

sk2001git committed:
No changes, committing in order to undo code during branch deletes

sk2001git committed:
Merge pull request #2 from sk2001git/branch-A-CodeQuality

Clean up code to better reflect code quality

sk2001git committed:
Merge branch 'master' into branch-A-Assertions

sk2001git committed:
Merge pull request #1 from sk2001git/branch-A-Assertions

Add assert to stop execution of programme given file corruption

sk2001git committed:
Merge branch 'master' of https://github.com/sk2001git/ip

sk2001git committed:
Add CLI github file for continuous integration on github

sk2001git committed:
Temporary save, error prone

sk2001git committed:
Add tag to the chatbot, ability to view tag, delete tag

Add feature to include tags now with task, tags are now stored alongside
with the task.

Used common logic for marking where the list is regenerated each time
after change of tags, can be possibly optimize later on.

In this case, design decision was to have a separate list for showing
tags that did not show their timings if they do have. This is to
prevent cluttering and better user experience for the user

Tested the launcher for functionality. Fixed data corruption problem
earlier where storage containing information would crash. Problem was
noted to be out of bounds error due to wrong checking logic

sk2001git committed:
Merge branch 'branch-C-Tagging'

sk2001git committed:
SLAP parse method in order to reach single level abstraction

sk2001git committed:
Remove gui start method, since its being loaded by fxml files

sk2001git committed:
Change header comments that were flagged for not following standard

sk2001git committed:
Clean up more documentation

sk2001git committed:
Revise Code Quality and Checkstyle

sk2001git committed:
Change gui slightly as well as images used due to copyright

sk2001git committed:
Merge branch 'branch-CodeQualityRevision'

sk2001git committed:
Add ui screenshot to represent product

sk2001git committed:
Squash some unintended output and update README.md

sk2001git committed:
Update test cases, assert in production code has cause invalid test

sk2001git committed:
Checkstyle amendments

sk2001git committed:
Update test text file

sk2001git committed:
Fixes to make result more consistent

sk2001git committed:
Merge branch 'branch-A-ErrorFixing'

sk2001git committed:
Remove system print ln files for storage for successful case

sk2001git committed:
Fix errorneous command

sk2001git committed:
Amend message for better gui

sk2001git committed:
Minor bug fix

sk2001git committed:
Update readme to include quickstart
